Laden Sie den Datensatz mit pandas

In [1]:
#IMPOPRT
import os

def start_workspace():
    curr = os.getcwd()
    while curr != os.path.dirname(curr):
        potential_ini = os.path.join(curr, 'src', 'ini.py')
        if os.path.exists(potential_ini):
            get_ipython().run_line_magic('run', f'-i "{potential_ini}"')
            return True
        curr = os.path.dirname(curr)
    return False

if not start_workspace():
    print("❌ Konnte 'src/ini.py' nicht finden. Prüfe die Ordnerstruktur!")

%load_ext autoreload
%autoreload 2

✅ Universal workspace ready!
📍 Main folder: /Users/cristallagus/Desktop/GitHub/weebet/_onboarding_data
📦 Modules loaded: pre_eda_words, pandas_deepseek_agent, pre_eda_data, translator


In [2]:
# DATA LOAD
df = load_data("daily_tasks/day_2/day_2_datasets/high-school-directory.csv")

In [4]:
# FULL DATA ANLYSIS
full_data_analysis(df)

********** CONSOLIDATED DATA QUALITY ANALYSIS **********
                                                                                                                        -->  -->  -->  
Shape (Rows, Columns): (435, 69) | Duplicate Rows: 0
--------------------------------------------------
   Datentyp    Semantischer_Typ  Einzigartige_Werte  Kardinalität(%)                            Spalte  Duplicate  NaN  NaN(%)           Min  Lower_Fence      25% (Q1)        Median         StdDev      75% (Q3)   Upper_Fence  Max/100%(Q4) Skewness Ausreißer (IQR) Ausreißer (StdDev)
0    object              object                 435           100.00                               dbn          0    0    0.00             -            -             -             -              -             -             -             -        -               -                  -
1    object     Text (Freitext)                 435           100.00                       school_name          0    0    0.00           

 1. = Type der spalten grade_span_min, ist float aber müste int sein. postcode ist als object zu behandeln. start_time & end_time ist eine Zeitspalte 
 2. = expgrade_span_min & expgrade_span_max: Über 92 % Fehlwerte (NaN). Diese Spalten sind in ihrem aktuellen Zustand statistisch kaum verwertbar. campus_name: 218 NaNs. nur etwa die Hälfte der Schulen Teil eines Campus-Systems ist. Dies muss bei Vergleichen zwischen "Single-Building" und "Campus" Schulen berücksichtigt werden.
 3. = subway: Fast 18 % Fehlwerte. Könnte die Analyse der Erreichbarkeit von Schulen verzerren.
 4. = state_code: Hat nur einen einzigen einzigartigen Wert (Kardinalität 0,23 %).
 5. = grade_span_max: Weist eine extrem hohe negative Schiefe (-4,54) auf. Mehrheit denselben Endwert hat (12), nur sehr wenige Schulen davon abweichen.

    - Bereinigung: Entfernen der Spalte state_code.
 - a. Imputation oder Drop: Entscheidung treffen, ob expgrade-Spalten aufgrund der 92 % Fehlwerte komplett gelöscht werden sollen.
 - b.Validierung: Prüfung der 8 doppelten Telefonnummern, um sicherzustellen, dass es sich nicht doch um versteckte Duplikate bei den Schulen handelt.

---

 1. = The data type of the `grade_span_min` column is `float`, but it should be `int`. `postcode` should be treated as an object. `start_time` and `end_time` are time columns. 
 2. = `expgrade_span_min` and `expgrade_span_max`: Over 92% of the values are missing (NaN). In their current state, these columns are statistically almost unusable. campus_name: 218 NaNs. Only about half of the schools are part of a campus system. This must be taken into account when comparing “single-building” and “campus” schools.
 3. = subway: Nearly 18% missing values. Could skew the analysis of school accessibility.
 4. = state_code: Has only a single unique value (cardinality 0.23%).
 5. = grade_span_max: Exhibits extremely high negative skewness (-4.54). The majority have the same final value (12), with only very few schools deviating from it.

    - Cleaning: Remove the state_code column.
 - a. Imputation or Drop: Decide whether to completely delete the expgrade columns due to the 92% missing values.
 - b. Validation: Check the 8 duplicate phone numbers to ensure they are not actually hidden duplicates among the schools.

In [6]:
#FULL WORDS ANALYSIS
full_words_analysis(df)

🚀 STARTING FULL COMPARISON ANALYSIS (52 Columns)
📊 1. STRUCTURE & QUALITY


,NaNs,Unique,Kardinalität %,Dom. Sprache,Sonderzeichen
Spalte,,,,,
dbn,0,435,100.0,N/A,0
school_name,0,435,100.0,en,60
borough,0,5,1.1,N/A,0
building_code,0,256,58.9,N/A,0
phone_number,0,427,98.2,N/A,870
fax_number,12,422,97.0,N/A,846
bus,1,235,54.0,N/A,115
subway,77,186,42.8,en,673
primary_address_line_1,0,258,59.3,en,84


🧠 2. CONTENT & SEMANTIC DEPTH


,Lex. Diversity,Stopword-Last,Top 5 Tokens,Zahlen im Text
Spalte,,,,
dbn,1.00,0.00,"27q260, 21k559, 16k393, 08x305, 03m485",435
school_name,1.00,0.13,"school, high, for, academy, and",6
borough,1.00,0.00,"brooklyn, bronx, manhattan, queens, staten",0
building_code,1.00,0.00,"x450, x405, m490, x410, x435",435
phone_number,1.00,0.00,"718-381-7100, 212-927-1841, 718-904-4200, 718-387-2800, 718-410-4242",435
fax_number,0.97,0.00,"212-674-8021, 718-471-2890, 718-333-7775, 718-574-2821, 718-824-1663",423
bus,1.00,0.00,"bx15,, m15,, m5,, m101,, bx17,",434
subway,0.73,0.01,"to, ;, st, ave, 2,",286
primary_address_line_1,1.00,0.00,"avenue, street, east, west, road",435


📏 3. LENGTH STATISTICS (RAW DATA)


,Ø Zeichen,Max Zeichen,Ø Wörter,Max Wörter
Spalte,,,,
dbn,6.0,6,1.0,1
school_name,36.9,79,5.5,12
borough,7.2,13,1.0,2
building_code,4.0,4,1.0,1
phone_number,12.0,14,1.0,1
fax_number,11.7,12,1.0,1
bus,37.7,141,7.4,29
subway,41.3,218,10.7,52
primary_address_line_1,18.9,30,3.3,5


⚠️ 4. ANOMALY BOARD (MANCOS)


,Leere Texte,Short (<5),Emojis,Shortforms,Shouting,Duplikate
Spalte,,,,,,
dbn,0,0,0,0,0,0
school_name,0,0,2,5,8,0
borough,0,0,0,0,0,430
building_code,0,435,0,0,0,179
phone_number,0,0,0,0,0,8
fax_number,12,12,0,0,0,12
bus,1,2,0,0,279,199
subway,77,77,0,8,0,248
primary_address_line_1,0,0,0,1,0,177


📊 1. STRUCTURE & QUALITY
- NaN-Raten:
Prioritäts-Spalten: Ab priority04 bis priority10 sind fast alle Werte NaN (über 90 % bis 99 %). Diese Spalten enthalten kaum Informationen.
Spezialisierte Programme: online_language_courses (362 NaNs) und psal_sports_coed (288 NaNs) sind sehr lückenhaft.
Partnerschaften: Während partner_highered relativ gut gefüllt ist, fehlen bei partner_financial (363 NaNs) und partner_corporate (237 NaNs) weite Teile der Daten.
- Sonderzeichen:
overview_paragraph & program_highlights 1.500 bis 2.500 Sonderzeiche
se_services & state_code Beide haben nur einen einzigen einzigartigen Wert. Das bedeutet, jede Zeile enthält dieselbe Information. Diese Spalten tragen nichts zur Unterscheidung der Schulen bei und können gelöscht werden.

**ell_programs: Nur 10 einzigartige Werte bei 435 Zeilen – hier herrscht eine sehr geringe Varianz.**

---

📊 1. STRUCTURE & QUALITY
- NaN rates:
Priority columns: From priority04 to priority10, almost all values are NaN (over 90% to 99%). These columns contain very little information.
Specialized programs: online_language_courses (362 NaNs) and psal_sports_coed (288 NaNs) are very incomplete.
Partnerships: While partner_highered is relatively well-filled, large portions of the data are missing in partner_financial (363 NaNs) and partner_corporate (237 NaNs).
- Special characters:
overview_paragraph & program_highlights 1,500 to 2,500 special characters
se_services & state_code Both have only a single unique value. This means every row contains the same information. These columns do not contribute to distinguishing between schools and can be deleted.

**ell_programs: Only 10 unique values across 435 rows – there is very little variation here.**


Translated with DeepL.com (free version)

🧠 2. CONTENT & SEMANTIC DEPTH

Stopword-Last in Kernspalten
 - overview_paragraph (21 %) & program_highlights (10 %) Stopword-Entfernung nötig bei ML
 - se_services (20 %) Dieser Wert ist extrem hoch, was darauf hindeutet, dass hier fast nur Standardsätze ohne spezifischen Informationsgehalt stehen.

Die Spalten für Sportarten und Partnerschaften zeigen eine inkonsistente Tokenisierung

Redundanz in den Prioritäts-Regeln
 - Die Textanalyse der priority-Spalten zeigt, dass sie fast identische Phrasen enthalten ("to", "residents", "priority", "then").kaum noch individuelle Informationen enthalten, sondern lediglich repetitive Standardtexte oder leere Worthülsen sind.

Sprachliche Dominanz
- alle analysierten Felder sind als "en" (Englisch) klassifiziert.

---

🧠 2. CONTENT & SEMANTIC DEPTH

High stopword content in core columns
 - overview_paragraph (21%) & program_highlights (10%) Stopword removal required for ML
 - se_services (20%) This value is extremely high, suggesting that these columns contain almost exclusively boilerplate text with no specific informational content.

The columns for sports and partnerships show inconsistent tokenization

Redundancy in the priority rules
 - Text analysis of the priority columns shows that they contain nearly identical phrases (“to,” “residents,” “priority,” “then”). They contain hardly any individual information, but are merely repetitive standard text or empty phrases.

Linguistic dominance
- All analyzed fields are classified as “en” (English).


📏 3. LENGTH STATISTICS (RAW DATA)

addtl_info1: 
- Diese Spalte ist mit durchschnittlich 23 Wörtern (Max. 168) überraschend inhaltsreich und scheint oft wichtigere Details zu enthalten als die spezifischen priority-Felder.

Normalisierung: 
- Da die start_time und end_time im Schnitt nur 6-7 Zeichen lang sind, sind sie perfekt für eine Umwandlung in ein einheitliches Zeitformat (HH:MM) geeignet.

---

📏 3. LENGTH STATISTICS (RAW DATA)

addtl_info1: 
- With an average of 23 words (max. 168), this column is surprisingly rich in content and often seems to contain more important details than the specific priority fields.

Normalization: 
- Since the start_time and end_time are on average only 6–7 characters long, they are perfectly suited for conversion to a uniform time format (HH:MM).

Bereinigen Sie die Spaltennamen (in Kleinbuchstaben umwandeln, Leerzeichen durch Kommas ersetzen _, Sonderzeichen entfernen).

Clean the column names (convert to lowercase, replace spaces with underscores _, remove special characters).

In [7]:
#🔎 Object Collums visualize
def analyze_semantic_and_row_integrity(df, max_unique=50):
    """
    Enhanced: Identifies object classes and checks for row-level integrity 
    including metadata detection and cardinality checks.
    """
    df_check = df
    potential_classes = []
    needs_manual_review = []
    
    print("SEMANTIC & ROW INTEGRITY ANALYSIS")
    
    for col in df_check.columns:
        # Preparation
        unique_vals = df_check[col].dropna().unique()
        unique_count = len(unique_vals)
        total_count = len(df_check[col])
        null_count = df_check[col].isnull().sum()
        nan_percentage = (null_count / total_count) * 100
        cardinality_ratio = (unique_count / total_count) * 100
        
        # 1. Basic Identification of Category/Object Class
        # Added: Check if it's not just a single repeated value (unique_count > 1)
        is_category = (1 < unique_count <= max_unique) or ('name' in col.lower() or 'dbn' in col.lower())
        
        if is_category:
            row_issues = []
            
            # A. Check for NaNs (Critical if > 90%)
            if null_count > 0:
                severity = "CRITICAL" if nan_percentage > 90 else "Warning"
                row_issues.append(f"{null_count} NaNs ({nan_percentage:.1f}%) [{severity}]")
            
            # B. Check for Mixed Types
            types_in_col = df_check[col].apply(type).unique()
            if len(types_in_col) > 1:
                row_issues.append(f"Mixed types: {list(types_in_col)}")
            
            # C. PLUS: Metadata Detection (URLs, Emails, Phone Numbers)
            if unique_count > 0:
                sample_str = str(unique_vals[0]).lower()
                patterns = {
                    'URL': r'http[s]?://',
                    'Email': r'[^@]+@[^@]+\.[^@]+',
                    'Phone': r'\d{3}-\d{3}-\d{4}'
                }
                for label, pattern in patterns.items():
                    if re.search(pattern, sample_str):
                        row_issues.append(f"Technical Metadata ({label}) detected")

            # D. PLUS: High Cardinality Check (Identifying hidden IDs)
            if cardinality_ratio > 95 and 'name' not in col.lower():
                row_issues.append(f"High Cardinality ({cardinality_ratio:.1f}%): Likely an ID, not a Class")

            # E. String Length Check
            if df_check[col].dtype == 'object' and unique_count <= max_unique:
                avg_len = pd.Series([len(str(x)) for x in unique_vals]).mean()
                if avg_len > 100:
                    row_issues.append("Text content too long for visualization labels")

            # Final Decision Logic
            if row_issues:
                needs_manual_review.append({
                    'column': col,
                    'unique_count': unique_count,
                    'issues': row_issues
                })
            else:
                potential_classes.append(col)

    return potential_classes, needs_manual_review

# EXECUTION
suggested_objects, review_list = analyze_semantic_and_row_integrity(df)

print("✅ SAFE OBJECT CLASSES (Clean rows & names):")
print(suggested_objects)

if review_list:
    print("⚠️ COLUMNS REQUIRING ROW-LEVEL CLEANING OR REVIEW:")
    for item in review_list:
        issues_str = " | ".join(item['issues'])
        print(f"- {item['column']} ({item['unique_count']} uniques): {issues_str}")

SEMANTIC & ROW INTEGRITY ANALYSIS
✅ SAFE OBJECT CLASSES (Clean rows & names):
['school_name', 'borough', 'grade_span_max', 'city', 'ell_programs', 'school_accessibility_description', 'number_programs', 'Borough Boundaries']
⚠️ COLUMNS REQUIRING ROW-LEVEL CLEANING OR REVIEW:
- dbn (435 uniques): High Cardinality (100.0%): Likely an ID, not a Class
- grade_span_min (3 uniques): 3 NaNs (0.7%) [Warning]
- expgrade_span_min (2 uniques): 404 NaNs (92.9%) [CRITICAL]
- expgrade_span_max (2 uniques): 402 NaNs (92.4%) [CRITICAL]
- campus_name (65 uniques): 218 NaNs (50.1%) [Warning] | Mixed types: [<class 'str'>, <class 'float'>]
- school_type (13 uniques): 331 NaNs (76.1%) [Warning] | Mixed types: [<class 'float'>, <class 'str'>]
- online_language_courses (40 uniques): 362 NaNs (83.2%) [Warning] | Mixed types: [<class 'str'>, <class 'float'>]
- start_time (38 uniques): 4 NaNs (0.9%) [Warning] | Mixed types: [<class 'str'>, <class 'float'>]
- priority03 (34 uniques): 192 NaNs (44.1%) [Warning] |

In [3]:
# Clean all column names
all_columns = df.columns

mapping = {c: re.sub(r'[^a-z0-9_]', '', c.lower().replace(" ", "_").replace("-", "_")) for c in all_columns}

# Before / After comparison table
comparison_df = pd.DataFrame({
    'Before': list(mapping.keys()),
    'After':  list(mapping.values())
})
comparison_df['Changed'] = comparison_df['Before'] != comparison_df['After']

print("📊 BEFORE / AFTER COMPARISON:")
print(comparison_df[comparison_df['Changed']].to_string(index=False))
print(f"\n🔄 {comparison_df['Changed'].sum()} column(s) were renamed.")

df.rename(columns=mapping, inplace=True)
print(f"\n✅ All {len(mapping)} column names cleaned.")

📊 BEFORE / AFTER COMPARISON:
                Before                  After  Changed
            Location 1             location_1     True
       Community Board        community_board     True
      Council District       council_district     True
          Census Tract           census_tract     True
                   BIN                    bin     True
                   BBL                    bbl     True
                   NTA                    nta     True
             Zip Codes              zip_codes     True
   Community Districts    community_districts     True
    Borough Boundaries     borough_boundaries     True
City Council Districts city_council_districts     True
      Police Precincts       police_precincts     True

🔄 12 column(s) were renamed.

✅ All 69 column names cleaned.


Filtern Sie den Datensatz so, dass er nur Schulen in Brooklyn enthält .

Filter the dataset so that it only contains schools in Brooklyn.

In [7]:
# Identify the correct column name (after our sanitization)
target_borough = 'brooklyn'
num_rows = 5
# Filter the DataFrame
df_brooklyn = df[df['borough'].str.lower() == target_borough].copy()
df_brooklyn = df_brooklyn.reset_index(drop=True)
# Validation
print("FILTERING REPORT")
print(f"Original schools: {len(df)}")
print(f"Schools in Brooklyn: {len(df_brooklyn)}")

# Show the first few rows of the new Brooklyn-only DataFrame
print(f"✅ First {num_rows} Schools in Brooklyn:")
print(df_brooklyn[['borough','dbn', 'school_name']].head(num_rows))

FILTERING REPORT
Original schools: 435
Schools in Brooklyn: 121
✅ First 5 Schools in Brooklyn:
    borough     dbn                                                school_name
0  Brooklyn  21K559                Life Academy High School for Film and Music
1  Brooklyn  16K393             Frederick Douglass Academy IV Secondary School
2  Brooklyn  17K122  Pathways in Technology Early College High School (P-Tech)
3  Brooklyn  23K644                             Eagle Academy for Young Men II
4  Brooklyn  15K462                                   Secondary School for Law


In [5]:
# Head Rows whit exclude
HEAD=2
exclude_cols = [
    #'dbn',
    'school_name',
    #'borough',
    #'building_code',
    'phone_number',
    'fax_number',
    #'grade_span_min',
    #'grade_span_max',
    'expgrade_span_min',
    'expgrade_span_max',
    'bus',
    'subway',
    'primary_address_line_1',
    'city',
    'state_code',
    'postcode',
    'website',
    #'total_students',
    'campus_name',
    'school_type',
    'overview_paragraph',
    'program_highlights',
    'language_classes',
    'advancedplacement_courses',
    'online_ap_courses',
    'online_language_courses',
    'extracurricular_activities',
    'psal_sports_boys',
    'psal_sports_girls',
    'psal_sports_coed',
    'school_sports',
    'partner_cbo',
    'partner_hospital',
    'partner_highered',
    'partner_cultural',
    'partner_nonprofit',
    'partner_corporate',
    'partner_financial',
    'partner_other',
    'addtl_info1',
    'addtl_info2',
    'start_time',
    'end_time',
    'se_services',
    'ell_programs',
    'school_accessibility_description',
    'number_programs',
    'priority01',
    'priority02',
    'priority03',
    'priority04',
    'priority05',
    'priority06',
    'priority07',
    'priority08',
    'priority09',
    'priority10',
    'Location 1',
    'community_board',
    'Council District',
    'Census Tract',
    'BIN',
    'BBL',
    'NTA',
    'Zip Codes',
    'Community Districts',
    'borough_boundaries',
    'City Council Districts',
    'Police Precincts'
]

# 2. Kompakte Bereinigungs-Funktion
def clean(c):
    return re.sub(r'[^a-z0-9_]', '', c.lower().replace(" ", "_").replace("-", "_"))

# 3. Transformation & Filterung (Inplace Umbenennung + List-Comprehension)
df.rename(columns={c: clean(c) for c in df.columns}, inplace=True)
cleaned_exclude = [clean(c) for c in exclude_cols]

# 4. Anzeige (Nur Spalten die NICHT in der Liste sind)
df[[c for c in df.columns if c not in cleaned_exclude]].head(HEAD).reset_index(drop=True)

,dbn,borough,building_code,grade_span_min,grade_span_max,total_students
0,27Q260,Queens,Q465,9.0,12,412.0
1,21K559,Brooklyn,K400,9.0,12,260.0


Beantworten Sie die folgenden Fragen

- Wie viele Schulen gibt es insgesamt in Brooklyn? (Verwenden Sie die Methode, die nur eindeutige Werte zählt. Es gibt auch eine Spalte mit einer eindeutigen Kennung für die Schulen.)

- Wie viele Schulen in Brooklyn bieten einen Einstieg in die 9. Klasse an ? (Hier muss ein Bereich angegeben werden, um die richtige Antwort zu erhalten; beispielsweise werden Schulen, die einen früheren Einstieg anbieten, nicht berücksichtigt, wenn nur „Eintritt in die 9. Klasse >= 9“ angegeben wird.)

Answer the following questions:

- How many schools are there in total in Brooklyn? (Use the method that counts only unique values. There is also a column with a unique identifier for the schools.)

- How many schools in Brooklyn offer entry into grade 9? (A range must be specified here to get the correct answer; for example, schools that offer an earlier entry will not be counted if only "entry into grade 9 >= 9" is specified.)

In [10]:
query('Wie viele Schulen gibt es insgesamt in Brooklyn? (Verwenden Sie die Methode, die nur eindeutige Werte zählt. Es gibt auch eine Spalte mit einer eindeutigen Kennung für die Schulen.)')

There are 121 unique schools in Brooklyn.



'There are 121 unique schools in Brooklyn.\n'

In [13]:
# Goal: Count unique schools in Brooklyn
# Filter: borough == 'Brooklyn'
def check_total_brooklyn_schools(df):
    brooklyn_data = df[df['borough'].str.contains('Brooklyn', case=False, na=False)]
    total_unique = brooklyn_data['dbn'].nunique()
    return total_unique

total_count = check_total_brooklyn_schools(df)
print(f"Total number of unique schools in Brooklyn: {total_count}")

Total number of unique schools in Brooklyn: 121


In [15]:
# Filters for borough 'Brooklyn' and an exact start at grade 9
def cross_check_brooklyn_9(df):
    """
    Direct calculation based on the data structure:
    Filters for borough 'Brooklyn' and an exact start at grade 9.
    """
    # 1. Filter: borough must be Brooklyn
    # 2. Filter: grade_span_min must be 9 (entry starting from grade 9 only)
    result_df = df[
        (df['borough'].str.contains('Brooklyn', case=False, na=False)) & 
        (df['grade_span_min'] >= 9)
    ]
    count = len(result_df)
    school_names = result_df['school_name'].tolist()
    return count, school_names

# Execution
count, names = cross_check_brooklyn_9(df)
print(f"Number of schools: {count}")
#print(f"Identified schools: {names}")

Number of schools: 98


Gruppieren und zusammenfassen :

- Anzahl der Schulen pro Stadtbezirk (wiederum einzelne Schulen)
- Durchschnittliche Schülerzahl pro Stadtbezirk
- Zusammenfassung, grade_span_maxgruppiert nach Stadtbezirk (verwenden Sie die Methode, die Ihnen grundlegende Statistiken zu dieser Spalte liefert)

Group and summarize:

- Number of schools per borough (individual schools again)
- Average number of students per borough
- Summary of grade_span_max grouped by borough (use the method that gives you basic statistics on this column)

In [6]:
query("in df zeige mir von quelle df Anzahl der Schulen pro bezirk,als Tabelle")

Bronx            118
Brooklyn         121
Manhattan        106
Queens            80
Staten Island     10
Name: borough, dtype: int64



'Bronx            118\nBrooklyn         121\nManhattan        106\nQueens            80\nStaten Island     10\nName: borough, dtype: int64\n'

In [18]:
# # Unique count per borough
def calculate_schools_per_borough(df):
    # Unique count per borough
    stats = df.groupby('borough')['dbn'].nunique().reset_index()
    stats.columns = ['Borough', 'Number_of_Schools']
    
    # Sort for better overview
    stats = stats.sort_values(by='Number_of_Schools', ascending=False)
    
    # Validation: sum must equal 435 (according to your shape report)
    total_check = stats['Number_of_Schools'].sum()
    
    return stats, total_check

boroughs_df, total_sum = calculate_schools_per_borough(df)
print(boroughs_df.to_string(index=False))
print(f"\nTotal sum verified: {total_sum} (Expected: 435)")

      Borough  Number_of_Schools
     Brooklyn                121
        Bronx                118
    Manhattan                106
       Queens                 80
Staten Island                 10

Total sum verified: 435 (Expected: 435)


In [11]:
query("from df, Average number of students per borough")

703.8427


703.8427

In [20]:
# Grouping and calculating the mean
def student_stats_per_borough(df):
    # Grouping and calculating the mean
    stats = df.groupby('borough')['total_students'].mean().round(2).reset_index()
    stats.columns = ['Borough', 'Average_Student_Count']
    
    # Sort by size
    stats = stats.sort_values(by='Average_Student_Count', ascending=False)
    
    return stats

# Execution
avg_students_df = student_stats_per_borough(df)
print("Average student count per borough:")
print(avg_students_df.to_string(index=False))

Average student count per borough:
      Borough  Average_Student_Count
Staten Island                1847.50
       Queens                1046.64
     Brooklyn                 699.13
    Manhattan                 589.83
        Bronx                 490.41


In [6]:
query("Summary of grade_span_max grouped by borough (use the method that gives you basic statistics on this column)")

Summary of grade_span_max grouped by borough:

               count       mean       std   min   25%   50%   75%   max
borough                                                                
Bronx          118.0  11.906780  0.369506   9.0  12.0  12.0  12.0  12.0
Brooklyn       121.0  11.933884  0.381566   9.0  12.0  12.0  12.0  12.0
Manhattan      106.0  11.877358  0.472135   9.0  12.0  12.0  12.0  12.0
Queens          80.0  11.825000  0.497462  10.0  12.0  12.0  12.0  12.0
Staten Island   10.0  12.000000  0.000000  12.0  12.0  12.0  12.0  12.0

Average student count per borough:
borough
Bronx             490.406780
Brooklyn          699.134454
Manhattan         589.825243
Queens           1046.644737
Staten Island    1847.500000
Name: total_students, dtype: float64



'Summary of grade_span_max grouped by borough:\n\n               count       mean       std   min   25%   50%   75%   max\nborough                                                                \nBronx          118.0  11.906780  0.369506   9.0  12.0  12.0  12.0  12.0\nBrooklyn       121.0  11.933884  0.381566   9.0  12.0  12.0  12.0  12.0\nManhattan      106.0  11.877358  0.472135   9.0  12.0  12.0  12.0  12.0\nQueens          80.0  11.825000  0.497462  10.0  12.0  12.0  12.0  12.0\nStaten Island   10.0  12.000000  0.000000  12.0  12.0  12.0  12.0  12.0\n\nAverage student count per borough:\nborough\nBronx             490.406780\nBrooklyn          699.134454\nManhattan         589.825243\nQueens           1046.644737\nStaten Island    1847.500000\nName: total_students, dtype: float64\n'

In [22]:
# Statistical summary of grade_span_max by borough
def grade_summary_per_borough(df):
    # The describe() method returns count, mean, std, min, 25%, 50%, 75%, max
    summary = df.groupby('borough')['grade_span_max'].describe()
    
    # Optional: Rename columns for English output
    summary.columns = ['Count', 'Mean', 'StdDev', 'Min', '25%', 'Median', '75%', 'Max']
    
    return summary

# Execution
grade_stats = grade_summary_per_borough(df)
print("Statistical summary of grade_span_max by borough:")
print(grade_stats)

Statistical summary of grade_span_max by borough:
               Count       Mean    StdDev   Min   25%  Median   75%   Max
borough                                                                  
Bronx          118.0  11.906780  0.369506   9.0  12.0    12.0  12.0  12.0
Brooklyn       121.0  11.933884  0.381566   9.0  12.0    12.0  12.0  12.0
Manhattan      106.0  11.877358  0.472135   9.0  12.0    12.0  12.0  12.0
Queens          80.0  11.825000  0.497462  10.0  12.0    12.0  12.0  12.0
Staten Island   10.0  12.000000  0.000000  12.0  12.0    12.0  12.0  12.0


Visualisierungen erstellen :

- Balkendiagramm: Anzahl der Schulen pro Stadtbezirk
- Notieren Sie am Ende des Notizbuchs 2–3 wichtige Erkenntnisse aus Ihren Ergebnissen.

Create visualizations:

- Bar chart: Number of schools per borough
- Note 2–3 key insights from your results at the end of the notebook.

In [4]:
query('quelle df, plot Balkendiagramm: Anzahl der Schulen school_name  pro Stadtbezirk borough')

''

In [17]:
# Aggregation Visualization Count of Schools Across Boroughs
def plot_schools_per_borough(df):
    # Aggregation
    counts = df.groupby('borough')['dbn'].nunique().sort_values(ascending=False).reset_index()
    counts.columns = ['Borough', 'Count']
    
    # Visualization
    fig = px.bar(counts,
                 x='Borough',
                 y='Count',
                 title='Distribution of Schools Across Boroughs',
                 text='Count',
                 template='plotly_white')
    
    fig.update_traces(textposition='outside')
    fig.show()

# Call
plot_schools_per_borough(df)

Einige Kernpunkte oder eine zusammenfassende Schlussfolgerung//Notieren Sie am Ende des Notizbuchs 2–3 wichtige Erkenntnisse aus Ihren Ergebnissen.

A few bullet points or a paragraph of conclusions//Write 2–3 key insights based on your findings at the end of the notebook

- Regionale Verteilung: Von den 435 Schulen entfällt die Mehrheit auf Brooklyn (121) und die Bronx (118), während Staten Island mit nur 10 Schulen die geringste Dichte aufweist.

- Schülerkapazität: Es herrscht ein starkes Gefälle bei der Schulgröße. Staten Island führt mit durchschnittlich 1.847,5 Schülern pro Schule, während die Bronx mit nur 490,4 Schülern die kleinsten Standorte hat.

- Strukturelle Einheitlichkeit: Über alle Bezirke hinweg liegt die Abschlussklasse im Median konstant bei 12, was den Fokus des Datensatzes auf High Schools mit Abschluss nach der 12. Klasse bestätigt.

Erstelle einen Pull Request in deinen Branch ( firstname-lastname)